In [1]:
import sys
import os
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# 현재 노트북 파일의 경로를 기준으로 프로젝트 루트 경로를 계산
project_root = os.path.abspath(os.path.join(os.getcwd(), '..', '..', '..', '..'))

# sys.path에 프로젝트 루트 경로가 없으면 추가
if project_root not in sys.path:
  sys.path.append(project_root)


In [2]:
from src.data.db_handler import DBHandler

db_handler = DBHandler(db_name="data_lake")

df = db_handler.fetch_data(table_name="kr_stock_basic_info")

column_index = df.columns

df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 70 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   id                         200 non-null    int64              
 1   ticker                     200 non-null    object             
 2   pdno                       200 non-null    object             
 3   prdt_type_cd               200 non-null    object             
 4   mket_id_cd                 200 non-null    object             
 5   scty_grp_id_cd             200 non-null    object             
 6   excg_dvsn_cd               200 non-null    object             
 7   setl_mmdd                  200 non-null    object             
 8   lstg_stqt                  200 non-null    object             
 9   lstg_cptl_amt              200 non-null    object             
 10  cpta                       200 non-null    object             
 11  papr  

id  ticker          pdno prdt_type_cd mket_id_cd scty_grp_id_cd  \
0   1  005930  00000A005930          300        STK             ST   
1   2  000660  00000A000660          300        STK             ST   
2   3  373220  00000A373220          300        STK             ST   
3   4  207940  00000A207940          300        STK             ST   
4   5  012450  00000A012450          300        STK             ST   

  excg_dvsn_cd setl_mmdd   lstg_stqt lstg_cptl_amt           cpta  papr  \
0           02        12  5919637922             0   778046685000   100   
1           02        12   728002365             0  3657652050000  5000   
2           02      1231   234000000             0   117000000000   500   
3           02      1231    71174000             0   177935000000  2500   
4           02      1200    51563401  427500000000   270317005000  5000   

  issu_pric kospi200_item_yn scts_mket_lstg_dt scts_mket_lstg_abol_dt  \
0         0                Y          19750611                          
1      5000                Y          19961226                          
2    300000                Y          20220127                          
3    136000                Y          20161110                          
4         0                Y          19870527                          

  kosdaq_mket_lstg_dt kosdaq_mket_lstg_abol_dt frbd_mket_lstg_dt  \
0                                                       19750611   
1                                                       19961226   
2                                                                  
3                                                                  
4                                                                  

  frbd_mket_lstg_abol_dt reits_kind_cd etf_dvsn_cd oilf_fund_yn  \
0                                                0            N   
1                                                0            N   
2                                                0            N   
3                                                0            N   
4                                                0            N   

  idx_bztp_lcls_cd idx_bztp_mcls_cd idx_bztp_scls_cd stck_kind_cd  \
0              002              013              013          101   
1              002              013              013          101   
2                                                             101   
3                                                             101   
4              002              014              014          101   

  mfnd_opng_dt mfnd_end_dt dpsi_erlm_cncl_dt etf_cu_qty     prdt_name  \
0                                                     0       삼성전자보통주   
1                                                     0   에스케이하이닉스보통주   
2                                                     0   LG에너지솔루션보통주   
3                                                     0   삼성바이오로직스보통주   
4                                                     0  한화에어로스페이스보통주   

   prdt_name120 prdt_abrv_name      std_pdno       prdt_eng_name  \
0       삼성전자보통주           삼성전자  KR7005930003  SamsungElectronics   
1   에스케이하이닉스보통주         SK하이닉스  KR7000660001            SK hynix   
2   LG에너지솔루션보통주       LG에너지솔루션  KR7373220003  LG Energy Solution   
3   삼성바이오로직스보통주       삼성바이오로직스  KR7207940008   SAMSUNG BIOLOGICS   
4  한화에어로스페이스보통주      한화에어로스페이스  KR7012450003    HANWHA AEROSPACE   

     prdt_eng_name120  prdt_eng_abrv_name dpsi_aptm_erlm_yn etf_txtn_type_cd  \
0  SamsungElectronics         SamsungElec                 Y               00   
1            SK hynix            SK hynix                 Y               00   
2  LG Energy Solution  LG Energy Solution                 Y               00   
3   SAMSUNG BIOLOGICS   SAMSUNG BIOLOGICS                 Y               00   
4    HANWHA AEROSPACE    HANWHA AEROSPACE                 Y               00   

  etf_type_cd lstg_abol_dt nwst_odst_dvsn_cd sbst_pric thco_sbst_pric  \
0                                          1     60570         

In [3]:
final_cols = column_index[1:]

In [4]:
for col in final_cols:
  df[col] .replace(r'^\s*$', np.nan, regex=True, inplace=True)
  df[col].replace(['0', 0], np.nan, inplace=True)

missing_value_counts = df[final_cols].isnull().sum()

missing_value_counts

ticker                         0
pdno                           0
prdt_type_cd                   0
mket_id_cd                     0
scty_grp_id_cd                 0
excg_dvsn_cd                   0
setl_mmdd                      0
lstg_stqt                      0
lstg_cptl_amt                120
cpta                           0
papr                           0
issu_pric                    100
kospi200_item_yn               0
scts_mket_lstg_dt              0
scts_mket_lstg_abol_dt       200
kosdaq_mket_lstg_dt          200
kosdaq_mket_lstg_abol_dt     200
frbd_mket_lstg_dt            191
frbd_mket_lstg_abol_dt       200
reits_kind_cd                200
etf_dvsn_cd                  200
oilf_fund_yn                   0
idx_bztp_lcls_cd              49
idx_bztp_mcls_cd              62
idx_bztp_scls_cd              62
stck_kind_cd                  13
mfnd_opng_dt                 200
mfnd_end_dt                  200
dpsi_erlm_cncl_dt            165
etf_cu_qty                   200
prdt_name 

In [5]:
missing_cols = missing_value_counts[missing_value_counts >= 100].index.to_list()

final_cols = [col for col in final_cols if col not in missing_cols]

print(f"결측 컬럼 : {missing_cols}")

print(f"[원본 컬럼 : {len(column_index)-1}] - [결측 컬럼 : {len(missing_cols)}] = [후보 컬럼 : {len(final_cols)}]")

결측 컬럼 : ['lstg_cptl_amt', 'issu_pric', 'scts_mket_lstg_abol_dt', 'kosdaq_mket_lstg_dt', 'kosdaq_mket_lstg_abol_dt', 'frbd_mket_lstg_dt', 'frbd_mket_lstg_abol_dt', 'reits_kind_cd', 'etf_dvsn_cd', 'mfnd_opng_dt', 'mfnd_end_dt', 'dpsi_erlm_cncl_dt', 'etf_cu_qty', 'etf_type_cd', 'lstg_abol_dt', 'crfd_item_yn', 'etf_chas_erng_rt_dbnb', 'lstg_rqsr_issu_istt_cd', 'lstg_rqsr_item_cd', 'trst_istt_issu_istt_cd']
[원본 컬럼 : 69] - [결측 컬럼 : 20] = [후보 컬럼 : 49]


In [6]:
nunique_counts = df[final_cols].nunique()

unique_cols = nunique_counts[nunique_counts == 1].index.to_list()

final_cols = [col for col in final_cols if col not in unique_cols]

print(f"유니크 컬럼 : {unique_cols}")

print(f"[원본 컬럼 : {len(column_index)-1}] - [결측 컬럼 : {len(missing_cols)}] - [유니크 컬럼 : {len(unique_cols)}] = [후보 컬럼 : {len(final_cols)}]")

유니크 컬럼 : ['prdt_type_cd', 'mket_id_cd', 'scty_grp_id_cd', 'excg_dvsn_cd', 'kospi200_item_yn', 'oilf_fund_yn', 'stck_kind_cd', 'etf_txtn_type_cd', 'nwst_odst_dvsn_cd', 'thco_sbst_pric_chng_dt', 'tr_stop_yn', 'admn_item_yn', 'clpr_chng_dt', 'elec_scty_yn', 'etf_etn_ivst_heed_item_yn', 'cptt_trad_tr_psbl_yn', 'nxt_tr_stop_yn', 'updated_at']
[원본 컬럼 : 69] - [결측 컬럼 : 20] - [유니크 컬럼 : 18] = [후보 컬럼 : 31]


In [7]:
df[final_cols].head()

ticker          pdno setl_mmdd   lstg_stqt           cpta  papr  \
0  005930  00000A005930        12  5919637922   778046685000   100   
1  000660  00000A000660        12   728002365  3657652050000  5000   
2  373220  00000A373220      1231   234000000   117000000000   500   
3  207940  00000A207940      1231    71174000   177935000000  2500   
4  012450  00000A012450      1200    51563401   270317005000  5000   

  scts_mket_lstg_dt idx_bztp_lcls_cd idx_bztp_mcls_cd idx_bztp_scls_cd  \
0          19750611              002              013              013   
1          19961226              002              013              013   
2          20220127              NaN              NaN              NaN   
3          20161110              NaN              NaN              NaN   
4          19870527              002              014              014   

      prdt_name  prdt_name120 prdt_abrv_name      std_pdno  \
0       삼성전자보통주       삼성전자보통주           삼성전자  KR7005930003   
1   에스케이하이닉스보통주   에스케이하이닉스보통주         SK하이닉스  KR7000660001   
2   LG에너지솔루션보통주   LG에너지솔루션보통주       LG에너지솔루션  KR7373220003   
3   삼성바이오로직스보통주   삼성바이오로직스보통주       삼성바이오로직스  KR7207940008   
4  한화에어로스페이스보통주  한화에어로스페이스보통주      한화에어로스페이스  KR7012450003   

        prdt_eng_name    prdt_eng_name120  prdt_eng_abrv_name  \
0  SamsungElectronics  SamsungElectronics         SamsungElec   
1            SK hynix            SK hynix            SK hynix   
2  LG Energy Solution  LG Energy Solution  LG Energy Solution   
3   SAMSUNG BIOLOGICS   SAMSUNG BIOLOGICS   SAMSUNG BIOLOGICS   
4    HANWHA AEROSPACE    HANWHA AEROSPACE    HANWHA AEROSPACE   

  dpsi_aptm_erlm_yn sbst_pric thco_sbst_pric thdt_clpr bfdy_clpr  \
0                 Y     60570          60570     79700     79700   
1                 Y    254160         254160    353000    353000   
2                 Y    265620         265620    349500    349500   
3                 Y    798720         798720   1024000   1024000   
4                 Y    654080         654080   1022000   1022000   

  std_idst_clsf_cd  std_idst_clsf_cd_name idx_bztp_lcls_cd_name  \
0           032604         통신 및 방송 장비 제조업               시가총액규모대   
1           032601                반도체 제조업               시가총액규모대   
2           032802         일차전지 및 축전지 제조업                   NaN   
3           032101  기초 의약물질 및 생물학적 제제 제조업                   NaN   
4           033103       항공기,우주선 및 부품 제조업               시가총액규모대   

  idx_bztp_mcls_cd_name idx_bztp_scls_cd_name ocr_no issu_istt_cd  \
0                 전기,전자                 전기,전자   0121        00593   
1                 전기,전자                 전기,전자   1147        00066   
2                   NaN                   NaN   0000        37322   
3                   NaN                   NaN   6590        20794   
4                  의료정밀                  의료정밀   0361        01245   

  stln_int_rt_dvsn_cd frnr_psnl_lmt_rt  
0                  01       0.00000000  
1                  01       0.00000000  
2                  01       0.00000000  
3                  01       0.00000000  
4                  01       0.00000000

| 컬럼명 | 한글컬럼명 | 중요도 | 이유 |
| :--- | :--- | :--- | :--- |
| `ticker` | 티커 (종목 코드) | **필수** | 데이터를 식별하고 시계열을 구성하는 가장 핵심적인 키 |
| `pdno` | 상품번호 | **불필요** | 종목을 식별하는 주요 키 중 하나, 하지만 'ticker'로 대체 가능 |
| `setl_mmdd` | 결산월일 | **하** | 특정 분기 분석 외에는 활용도가 낮고, `datetime` 타입이 아님 |
| `lstg_stqt` | 상장주식수 | **상** | 시가총액 계산 및 유동성 파악에 필요한 핵심 지표 |
| `cpta` | 자본금 | **상** | 기업의 규모를 나타내는 중요한 기본적 분석 지표 |
| `papr` | 액면가 | **중** | 기본적인 정보이나, 시장 가격 대비 중요도는 낮음 |
| `scts_mket_lstg_dt` | 유가증권시장 상장일자 | **상** | 종목의 역사와 데이터 시작점을 아는 데 중요함 |
| `idx_bztp_lcls_cd` | 지수업종 대분류코드 | **중** | 섹터별 분석을 위한 그룹핑 변수 |
| `idx_bztp_mcls_cd` | 지수업종 중분류코드 | **중** | 대분류보다 세분화된 섹터 분석 시 유용함 |
| `idx_bztp_scls_cd` | 지수업종 소분류코드 | **중** | 중분류보다 세분화된 섹터 분석 시 유용함 |
| `prdt_name` | 상품명 | **상** | 분석 결과 해석 및 시각화 시 가독성을 높여주는 핵심 정보 |
| `prdt_name120` | 상품명120 | **불필요** | `prdt_name`과 정보가 중복됨 |
| `prdt_abrv_name` | 상품 약어명 | **중** | 시각화나 표에서 공간이 부족할 때 유용하나, `prdt_name`으로 대체 가능 |
| `std_pdno` | 표준 상품번호 | **상** | 외부 데이터와 결합할 때 사용할 수 있는 표준 식별자, 하지만 'ticker'로 대체 가능 |
| `prdt_eng_name` | 상품 영문명 | **불필요** | 국내 시장 분석에서는 활용도가 거의 없으며, `prdt_name`과 중복됨 |
| `prdt_eng_name120` | 상품 영문명120 | **불필요** | `prdt_eng_name`과 정보가 중복됨 |
| `prdt_eng_abrv_name` | 상품 영문 약어명 | **불필요** | `prdt_eng_name`과 정보가 중복됨 |
| `dpsi_aptm_erlm_yn` | 예탁지정 등록여부 | **하** | 특정 제도 분석 외에는 활용도가 낮음 |
| `sbst_pric` | 대용가 | **중** | 리스크 관리 측면(담보 가치)에서 활용될 수 있는 지표 |
| `thco_sbst_pric` | 당사대용가 | **불필요** | `sbst_pric`과 정보가 유사하거나 특정 증권사에 종속된 정보일 가능성 |
| `thdt_clpr` | 당일종가 | **필수** | 변동성 및 수익률 계산의 가장 기본이 되는 가격 정보 |
| `bfdy_clpr` | 전일종가 | **필수** | 일별 수익률 계산에 반드시 필요한 정보 |
| `std_idst_clsf_cd` | 표준산업분류코드 | **상** | 산업별 특성을 분석하기 위한 핵심 그룹핑 변수 |
| `std_idst_clsf_cd_name`| 표준산업분류코드명 | **상** | `std_idst_clsf_cd`의 해석을 위해 함께 사용하는 것이 좋음 |
| `idx_bztp_lcls_cd_name`| 지수업종 대분류명 | **중** | `idx_bztp_lcls_cd`의 해석을 위해 사용 |
| `idx_bztp_mcls_cd_name`| 지수업종 중분류명 | **중** | `idx_bztp_mcls_cd`의 해석을 위해 사용 |
| `idx_bztp_scls_cd_name`| 지수업종 소분류명 | **중** | `idx_bztp_scls_cd`의 해석을 위해 사용 |
| `ocr_no` | OCR 번호 | **하** | 내부 관리용 번호일 가능성이 높아 분석적 가치가 낮음 |
| `issu_istt_cd` | 발행기관코드 | **상** | 발행 주체를 식별하는 중요한 정보 |
| `stln_int_rt_dvsn_cd` | 대주이자율 구분코드 | **하** | 공매도 등 특정 전략 분석이 아니면 활용도가 낮음 |
| `frnr_psnl_lmt_rt` | 외국인 개인 한도비율 | **중** | 수급 분석 및 이벤트 분석 시 활용될 수 있는 정보 |

In [8]:
final_cols = [
  'ticker',
  'lstg_stqt',
  'cpta',
  'papr',
  'scts_mket_lstg_dt',
  'idx_bztp_lcls_cd',
  'idx_bztp_mcls_cd',
  'idx_bztp_scls_cd',
  'prdt_name',
  'prdt_abrv_name',
  'std_pdno',
  'sbst_pric',
  'thdt_clpr',
  'bfdy_clpr',
  'std_idst_clsf_cd',
  'std_idst_clsf_cd_name',
  'idx_bztp_lcls_cd_name',
  'idx_bztp_mcls_cd_name',
  'idx_bztp_scls_cd_name',
  'issu_istt_cd',
  'frnr_psnl_lmt_rt'
]

In [9]:
df[final_cols].head()

ticker   lstg_stqt           cpta  papr scts_mket_lstg_dt idx_bztp_lcls_cd  \
0  005930  5919637922   778046685000   100          19750611              002   
1  000660   728002365  3657652050000  5000          19961226              002   
2  373220   234000000   117000000000   500          20220127              NaN   
3  207940    71174000   177935000000  2500          20161110              NaN   
4  012450    51563401   270317005000  5000          19870527              002   

  idx_bztp_mcls_cd idx_bztp_scls_cd     prdt_name prdt_abrv_name  \
0              013              013       삼성전자보통주           삼성전자   
1              013              013   에스케이하이닉스보통주         SK하이닉스   
2              NaN              NaN   LG에너지솔루션보통주       LG에너지솔루션   
3              NaN              NaN   삼성바이오로직스보통주       삼성바이오로직스   
4              014              014  한화에어로스페이스보통주      한화에어로스페이스   

       std_pdno sbst_pric thdt_clpr bfdy_clpr std_idst_clsf_cd  \
0  KR7005930003     60570     79700     79700           032604   
1  KR7000660001    254160    353000    353000           032601   
2  KR7373220003    265620    349500    349500           032802   
3  KR7207940008    798720   1024000   1024000           032101   
4  KR7012450003    654080   1022000   1022000           033103   

   std_idst_clsf_cd_name idx_bztp_lcls_cd_name idx_bztp_mcls_cd_name  \
0         통신 및 방송 장비 제조업               시가총액규모대                 전기,전자   
1                반도체 제조업               시가총액규모대                 전기,전자   
2         일차전지 및 축전지 제조업                   NaN                   NaN   
3  기초 의약물질 및 생물학적 제제 제조업                   NaN                   NaN   
4       항공기,우주선 및 부품 제조업               시가총액규모대                  의료정밀   

  idx_bztp_scls_cd_name issu_istt_cd frnr_psnl_lmt_rt  
0                 전기,전자        00593       0.00000000  
1                 전기,전자        00066       0.00000000  
2                   NaN        37322       0.00000000  
3                   NaN        20794       0.00000000  
4                  의료정밀        01245       0.00000000

In [10]:
final_df = df[final_cols].copy()

# --- 1단계: 날짜 타입 자동 변환 ---
remaining_object_cols = final_df.select_dtypes(include='object').columns.tolist()
for col in remaining_object_cols:
  original_valid_count = final_df[col].count()
  if original_valid_count == 0: continue
  
  for format in ['%Y%m', '%Y%m%d']:
    date_series = pd.to_datetime(final_df[col], format=format,errors='coerce')
    success_rate = date_series.count() / original_valid_count
    
    if success_rate > 0.95: # 95% 임계값
      final_df[col] = date_series
      break

# --- 2단계: 저비율 카테고리 자동 변환 ---
object_cols = final_df.select_dtypes(include='object').columns.tolist()
for col in object_cols:
  ratio = final_df[col].nunique() / len(final_df[col])
  if ratio < 0.05: # 5% 임계값
    final_df[col] = final_df[col].astype('category')


In [11]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ticker                 200 non-null    object        
 1   lstg_stqt              200 non-null    object        
 2   cpta                   200 non-null    object        
 3   papr                   200 non-null    category      
 4   scts_mket_lstg_dt      200 non-null    datetime64[ns]
 5   idx_bztp_lcls_cd       151 non-null    category      
 6   idx_bztp_mcls_cd       138 non-null    object        
 7   idx_bztp_scls_cd       138 non-null    object        
 8   prdt_name              200 non-null    object        
 9   prdt_abrv_name         200 non-null    object        
 10  std_pdno               200 non-null    object        
 11  sbst_pric              200 non-null    object        
 12  thdt_clpr              200 non-null    object        
 13  bfdy_

| 컬럼명 | 한글명 | 현재타입 | 적정타입 | 이유 |
| :--- | :--- | :--- | :--- | :--- |
| `ticker` | 티커 (종목 코드) | object | category | 데이터를 식별하고 그룹화하는 핵심 키이므로 category가 연산에 유리합니다. |
| `lstg_stqt` | 상장주식수 | object | int64 | 양적인 수치 정보를 나타내므로 숫자형(정수)으로 변환해야 합니다. |
| `cpta` | 자본금 | object | int64 | 기업 규모를 나타내는 양적 지표이므로 숫자형(정수)으로 변환해야 합니다. |
| `papr` | 액면가 | category | int64 | 금액을 나타내는 양적 지표이므로 숫자형(정수)으로 변환해야 합니다. |
| `scts_mket_lstg_dt` | 유가증권시장 상장일자 | datetime64[ns] | datetime64[ns] | 이미 올바른 날짜 타입으로 지정되어 있습니다. |
| `idx_bztp_lcls_cd` | 지수업종 대분류코드 | category | category | 이미 올바른 카테고리 타입으로 지정되어 있습니다. |
| `idx_bztp_mcls_cd` | 지수업종 중분류코드 | object | category | 반복되는 코드 값이므로 메모리 및 연산 효율을 위해 category가 적합합니다. |
| `idx_bztp_scls_cd` | 지수업종 소분류코드 | object | category | 반복되는 코드 값이므로 메모리 및 연산 효율을 위해 category가 적합합니다. |
| `prdt_name` | 상품명 | object | object | 고유값이 많아 카테고리 변환의 이점이 없으므로 object로 유지합니다. |
| `prdt_abrv_name` | 상품 약어명 | object | object | 고유값이 많아 카테고리 변환의 이점이 없으므로 object로 유지합니다. |
| `std_pdno` | 표준 상품번호 | object | category | 데이터를 식별하는 핵심 키이므로 category가 연산에 유리합니다. |
| `sbst_pric` | 대용가 | object | int64 | 금액을 나타내는 양적 지표이므로 숫자형(정수)으로 변환해야 합니다. |
| `thdt_clpr` | 당일종가 | object | float64 | 가격 정보를 나타내는 양적 지표이므로 숫자형(실수)으로 변환해야 합니다. |
| `bfdy_clpr` | 전일종가 | object | float64 | 가격 정보를 나타내는 양적 지표이므로 숫자형(실수)으로 변환해야 합니다. |
| `std_idst_clsf_cd` | 표준산업분류코드 | object | category | 반복되는 코드 값이므로 메모리 및 연산 효율을 위해 category가 적합합니다. |
| `std_idst_clsf_cd_name`| 표준산업분류코드명 | object | category | 분류 코드명으로 반복이 많으므로 category가 적합합니다. |
| `idx_bztp_lcls_cd_name`| 지수업종 대분류명 | category | category | 이미 올바른 카테고리 타입으로 지정되어 있습니다. |
| `idx_bztp_mcls_cd_name`| 지수업종 중분류명 | object | category | 분류 코드명으로 반복이 많으므로 category가 적합합니다. |
| `idx_bztp_scls_cd_name`| 지수업종 소분류명 | object | category | 분류 코드명으로 반복이 많으므로 category가 적합합니다. |
| `issu_istt_cd` | 발행기관코드 | object | category | 데이터를 식별하는 핵심 키이므로 category가 연산에 유리합니다. |
| `frnr_psnl_lmt_rt` | 외국인 개인 한도비율 | category | float64 | 비율을 나타내는 양적 지표이므로 숫자형(실수)으로 변환해야 합니다. |

In [12]:
# 수정할 타입을 딕셔너리로 관리
dtype_map = {
  'int64': ['lstg_stqt', 'cpta', 'papr', 'sbst_pric'],
  'float64': ['thdt_clpr', 'bfdy_clpr', 'frnr_psnl_lmt_rt'],
  'category': ['ticker', 'idx_bztp_mcls_cd', 'idx_bztp_scls_cd', 'std_pdno', 'std_idst_clsf_cd', 'std_idst_clsf_cd_name', 'idx_bztp_mcls_cd_name', 'idx_bztp_scls_cd_name', 'issu_istt_cd']
}

for dtype, cols in dtype_map.items():
  # DataFrame에 실제 존재하는 컬럼만 필터링
  valid_cols = [col for col in cols if col in final_df.columns]
  if not valid_cols:
    continue

  try:
    if dtype == 'Int64':
      for col in valid_cols:
        numeric_col = pd.to_numeric(final_df[col], errors='coerce')
        if not numeric_col.isnull().all():
          final_df[col] = numeric_col.astype('Int64')
    
    elif dtype == 'float64':
      final_df[valid_cols] = final_df[valid_cols].apply(pd.to_numeric, errors='coerce')

    elif dtype == 'datetime64':
      for col in valid_cols:
        final_df[col] = pd.to_datetime(final_df[col], errors='coerce')
    
    else: # 'category' 등 .astype()으로 처리 가능한 나머지 타입
      final_df[valid_cols] = final_df[valid_cols].astype(dtype)

  except (ValueError, TypeError) as e:
    print(f"⚠️ 경고: 컬럼 {valid_cols}을(를) '{dtype}'으로 변환 중 오류. 건너뜁니다. (에러: {e})")

In [13]:
final_df.head()

ticker   lstg_stqt           cpta  papr scts_mket_lstg_dt idx_bztp_lcls_cd  \
0  005930  5919637922   778046685000   100        1975-06-11              002   
1  000660   728002365  3657652050000  5000        1996-12-26              002   
2  373220   234000000   117000000000   500        2022-01-27              NaN   
3  207940    71174000   177935000000  2500        2016-11-10              NaN   
4  012450    51563401   270317005000  5000        1987-05-27              002   

  idx_bztp_mcls_cd idx_bztp_scls_cd     prdt_name prdt_abrv_name  \
0              013              013       삼성전자보통주           삼성전자   
1              013              013   에스케이하이닉스보통주         SK하이닉스   
2              NaN              NaN   LG에너지솔루션보통주       LG에너지솔루션   
3              NaN              NaN   삼성바이오로직스보통주       삼성바이오로직스   
4              014              014  한화에어로스페이스보통주      한화에어로스페이스   

       std_pdno  sbst_pric  thdt_clpr  bfdy_clpr std_idst_clsf_cd  \
0  KR7005930003      60570      79700      79700           032604   
1  KR7000660001     254160     353000     353000           032601   
2  KR7373220003     265620     349500     349500           032802   
3  KR7207940008     798720    1024000    1024000           032101   
4  KR7012450003     654080    1022000    1022000           033103   

   std_idst_clsf_cd_name idx_bztp_lcls_cd_name idx_bztp_mcls_cd_name  \
0         통신 및 방송 장비 제조업               시가총액규모대                 전기,전자   
1                반도체 제조업               시가총액규모대                 전기,전자   
2         일차전지 및 축전지 제조업                   NaN                   NaN   
3  기초 의약물질 및 생물학적 제제 제조업                   NaN                   NaN   
4       항공기,우주선 및 부품 제조업               시가총액규모대                  의료정밀   

  idx_bztp_scls_cd_name issu_istt_cd  frnr_psnl_lmt_rt  
0                 전기,전자        00593               0.0  
1                 전기,전자        00066               0.0  
2                   NaN        37322               0.0  
3                   NaN        20794               0.0  
4                  의료정밀        01245               0.0

In [14]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ticker                 200 non-null    category      
 1   lstg_stqt              200 non-null    int64         
 2   cpta                   200 non-null    int64         
 3   papr                   200 non-null    int64         
 4   scts_mket_lstg_dt      200 non-null    datetime64[ns]
 5   idx_bztp_lcls_cd       151 non-null    category      
 6   idx_bztp_mcls_cd       138 non-null    category      
 7   idx_bztp_scls_cd       138 non-null    category      
 8   prdt_name              200 non-null    object        
 9   prdt_abrv_name         200 non-null    object        
 10  std_pdno               200 non-null    category      
 11  sbst_pric              200 non-null    int64         
 12  thdt_clpr              200 non-null    int64         
 13  bfdy_